ir/src/main/scala/org/scalajs/core/ir/Printers.scala

```diff
@@ -40,50 +42,51 @@ object Printers {
         out.write(indentString, 0, indentMargin)
     }
 
-    @tailrec
-    protected final def printSeq[A](ls: List[A])(printelem: A => Unit)(
-        printsep: A => Unit): Unit = {
-      ls match {
-        case Nil =>
-        case x :: Nil =>
-          printelem(x)
-        case x :: rest =>
-          printelem(x)
-          printsep(x)
-          printSeq(rest)(printelem)(printsep)
-      }
-    }
-
-    protected def printColumn(ts: List[Any], start: String, sep: String,
+    protected final def printColumn(ts: List[Any], start: String, sep: String,
         end: String): Unit = {
-      print(start); indent(); println()
-      printSeq(ts) { x =>
-        print(x)
-      } { _ =>
-        print(sep)
-        println()
+      printString(start); indent(); println()
+      var rest = ts
+      while (rest.nonEmpty) {
+        printOne(rest.head)
+        rest = rest.tail
+        if (rest.nonEmpty) {
+          printString(sep)
+          println()
+        }
       }
-      undent(); println(); print(end)
+      undent(); println(); printString(end)
     }
 
-    protected def printRow(ts: List[Any], start: String, sep: String,
+    protected final def printRow(ts: List[Any], start: String, sep: String,
         end: String): Unit = {
-      print(start)
-      printSeq(ts) { x =>
-        print(x)
-      } { _ =>
-        print(sep)
+      printString(start)
+      var rest = ts
+      while (rest.nonEmpty) {
+        printOne(rest.head)
+        rest = rest.tail
+        if (rest.nonEmpty)
+          printString(sep)
       }
-      print(end)
+      printString(end)
     }
 
-    protected def printRow(ts: List[Any], sep: String): Unit =
+    protected final def printRow(ts: List[Any], sep: String): Unit =
       printRow(ts, "", sep, "")
 
-    protected def print(args: Any*): Unit =
-      args.foreach(printOne)
+    protected final def print(args: Any*): Unit = {
+      var i = 0
+      val len = args.length
+      while (i < len) {
+        printOne(args(i))
+        i += 1
+      }
+    }
 
     protected def printOne(arg: Any): Unit
+
+    protected def printString(s: String): Unit
+
+    protected def printChar(c: Int): Unit
```

Comment:
I would suggest overloads rather than using different names. Overloads are resolved at compile time, and they do not need "human" specialization at call site.